File: editDistance.py
#jiuzhang two sequence dp solution, very neat and clean approach
#for explaination, see spreadsheet on lesson 6
class Solution(object):
    def minDistance(self, word1, word2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        """
        m, n = len(word1), len(word2)
        dp = [[0] * (n+1) for _ in xrange(m+1)]
        for i in xrange(m + 1):
            for j in xrange(n + 1):
                if not i:
                    dp[i][j] = j
                    continue
                elif not j:
                    dp[i][j] = i
                    continue
                #delete tail, swap tail and add extra to tail
                dp[i][j] = min(dp[i-1][j] + 1, dp[i-1][j-1] + 1, dp[i][j-1] + 1)
                #when last char is the same, no one more step, only update [i-1][j-1]
                if word1[i-1] == word2[j-1]:
                    dp[i][j] = min(dp[i][j], dp[i-1][j-1])
        return dp[m][n]

File: editDistance.kt
class Solution {
// approach using mXn matrix
    fun minDistance(word1: String, word2: String): Int {
        val m = word1.length
        val n = word2.length
        val res = Array(m+1) { IntArray(n+1) {0}}
        for(i in 0 .. m){
            for(j in 0 .. n){
                // distance for empty a to non-empty b
                if(i == 0) res[i][j] = j
                // distance for non-empty a to empty b
                else if(j == 0) res[i][j] = i
                else {
                    //                   replace last char, add last char to a, delete last char from a
                    res[i][j] = intArrayOf(res[i-1][j-1] + 1, res[i][j-1] + 1, res[i-1][j] + 1).min()
                    // when last char of a == last char of b
                    if(word1[i-1] == word2[j-1]) res[i][j] = intArrayOf(res[i][j], res[i-1][j-1]).min()
                }
            }
        }
        return res[m][n]
    }
}
